tx_test.go

```diff
@@ -18,6 +21,105 @@ import (
 	"go.etcd.io/bbolt/internal/btesting"
 )
 
+func TestTx_Check_Page(t *testing.T) {
+	bucketKey := "testBucket"
+	db := btesting.MustCreateDBWithOption(t, &bolt.Options{PageSize: pageSize})
+	defer db.MustDeleteFile()
+
+	// arrange
+	if err := db.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists([]byte(bucketKey))
+		if err != nil {
+			t.Fatal(err)
+		}
+		// insert random data
+		for i := 0; i < 10; i++ {
+			generateSampleDataInBucket(t, b, 3)
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+	db.MustCheck()
+
+	// act : corrupt a leaf page
+	corruptLeafPageFromDB(t, db.Path())
+
+	// assert: expected an error after check
+	if err := db.View(func(tx *bolt.Tx) error {
+		for err := range tx.Check() {
+			fmt.Println(err)
+		}
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+	db.MustClose()
+}
+
+func corruptLeafPageFromDB(t testing.TB, dbPath string) {
+	_, hwm, err := guts_cli.ReadPageAndHWMSize(dbPath)
+	if err != nil {
+		t.Fatalf("error reading HWM from dbPath %v: %v", dbPath, err)
+	}
+
+	var victimPage *common.Page
+	var victimBuf []byte
+
+	// find first leaf page starting page after root bucket's page
+	for pageID := uint64(4); pageID < uint64(hwm); {
+		p, buf, err := guts_cli.ReadPage(dbPath, pageID)
+		if err != nil {
+			t.Fatalf("error reading page %v from dbPath %v: %v", pageID, dbPath, err)
+		}
+
+		if p.IsLeafPage() {
+			victimPage = p
+			victimBuf = buf
+			break
+		}
+	}
+
+	if victimPage == nil || victimBuf == nil {
+		t.Fatalf("no leaf page is found in dbPath %v", dbPath)
+	}
+
+	// corrupt page's type
+	buf := []byte("this is to corrupt the page")
+	copy(victimBuf[17:], buf)
```

Comment:
as discussed, i just dump some data into the page 

initially i used 

```
copy(victimBuf[:], buf)
```

however, the writing failed because the page header was corrupted and https://github.com/etcd-io/bbolt/blob/fdd0d1b8fd80c20d4f1c3db85bebe885f93c4256/tx_test.go#L91 checks the the size of the page before writing to disk 

so i started flushing random data into the page from byte `17` since the page header is `16` 

yet, the test case fails due to panic, i expected that `tx.Check()` reports the issues without panic 

please correct me if i am wrong

Reply:
cc @ahrtr @fuweid

Reply:
I think you can use f.WriteAt instead of calling `guts_cli.WritePage` since you already have victimPage.
So, IMO, it makes sense to use `copy(victimBuf[:]`

Reply:
> iiuc f.WriteAt directly manipulates the mmap

We can close db and manipulate db and reopen to check.

> also if you can run the test case locally, and check why it panics instead of reporting an error as expected ?

I did run it locally.

```bash
➜  bbolt git:(review-659) go test -v -run TestTx_Check_Page ./
seed: 95249
quick settings: count=5, items=1000, ksize=1024, vsize=1024
=== RUN   TestTx_Check_Page
    btesting.go:47: Opening bbolt DB at: /tmp/TestTx_Check_Page1621400540/001/db
panic: runtime error: slice bounds out of range [::2480444563] with length 268435455

goroutine 8 [running]:
go.etcd.io/bbolt/internal/common.UnsafeByteSlice(...)
        /home/fuwei/workspace/bbolt/internal/common/unsafe.go:26
go.etcd.io/bbolt/internal/common.(*leafPageElement).Key(...)
        /home/fuwei/workspace/bbolt/internal/common/page.go:301
go.etcd.io/bbolt.(*Tx).recursivelyCheckPageKeyOrderInternal(0xc000063cc8?, 0x4, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}, {0x0, 0x0, ...}, ...)
        /home/fuwei/workspace/bbolt/tx_check.go:163 +0x52c
go.etcd.io/bbolt.(*Tx).recursivelyCheckPageKeyOrder(...)
        /home/fuwei/workspace/bbolt/tx_check.go:129
go.etcd.io/bbolt.(*Tx).checkInvariantProperties(0xc0000e6540, 0x4, 0x619380?, 0x8249c0?, {0x6d0660?, 0x881aa0}, 0xc000082480)
        /home/fuwei/workspace/bbolt/tx_check.go:99 +0x14e
go.etcd.io/bbolt.(*Tx).recursivelyCheckBucket(0xc0000e6540, 0xc0000e6558, 0xc000063e90, 0xc000063ec0, {0x6d0660, 0x881aa0}, 0xc000082480)
        /home/fuwei/workspace/bbolt/tx_check.go:82 +0x73
go.etcd.io/bbolt.(*Tx).check(0xc0000e6540, {{0x6d0660?, 0x881aa0?}}, 0xc000050790?)
        /home/fuwei/workspace/bbolt/tx_check.go:64 +0x365
go.etcd.io/bbolt.(*Tx).Check.func1()
        /home/fuwei/workspace/bbolt/tx_check.go:33 +0x58
created by go.etcd.io/bbolt.(*Tx).Check in goroutine 6
        /home/fuwei/workspace/bbolt/tx_check.go:30 +0xea
FAIL    go.etcd.io/bbolt        0.042s
FAIL
```

The `copy(victimBuf[17:], buf)` is to change leaf layout. When tx_check uses unexpected `pos` and `size`, based on your content `this is to corrupt the page`, the element value is not correct.

```go
// leafPageElement represents a node on a leaf page.
type leafPageElement struct {
        flags uint32
        pos   uint32
        ksize uint32
        vsize uint32
}
```

first one's pos will be `s is`.

```
package main

import (
        "fmt"
        "unsafe"
)

func main() {
        str := "s is"
        var x = *(*uint32)(unsafe.Pointer((*byte)(unsafe.StringData(str))))
        fmt.Println(x)
}
```

Reply:
IMO, we can use invalid `leafPageElement` value to override original one. It would be easy to reason about why it panic.

Reply:
done